libs/ledger-live-common/src/families/cosmos/api/Cosmos.ts

```diff
@@ -229,49 +229,94 @@ export class CosmosAPI {
+    let txs: CosmosTx[] = [];
     try {
-      send = await this.getSendingTransactions(address, "ORDER_BY_DESC");
-      receive = await this.getReceivingTransactions(address, "ORDER_BY_DESC");
-    } catch (e) {
-      // FIXME: remove and always orderBy once pagination is implemented or every nodes are upgraded to 0.47
-      // See https://github.com/cosmos/cosmos-sdk/issues/15437
-      if ((e as Error).message.includes("ORDER_BY_DESC")) {
-        send = await this.getSendingTransactions(address);
-        receive = await this.getReceivingTransactions(address);
-      }
-    }
+      let total: number;
+      let previousTxCall: CosmosTx[] | null = null;
+      let paginationOffset = 0;
+
+      do {
+        const response: {
+          txs: CosmosTx[];
+          nextPageToken: string | undefined;
+          total: number;
+        } = await this.fetchTransactions(this.defaultEndpoint, filterOn, address, {
+          "pagination.limit": paginationSize,
+          "pagination.offset": paginationOffset,
+          "pagination.reverse": true,
+        });
 
-    return [...receive.data.tx_responses, ...send.data.tx_responses];
-  };
+        if (
+          previousTxCall &&
+          previousTxCall.length === txs.length &&
+          response.txs[0] &&
+          previousTxCall[0].txhash === response.txs[0].txhash
+        ) {
+          // Means we are getting the same thing... prevents an infinite loop
+          break;
+        }
 
-  private async getReceivingTransactions(
-    address: string,
-    orderBy?: "ORDER_BY_DESC" | "ORDER_BY_ASC",
-  ) {
-    return await network({
-      method: "GET",
-      url:
-        `${this.defaultEndpoint}/cosmos/tx/${this.version}/txs?events=` +
-        encodeURI(`transfer.recipient='${address}'`) +
-        (orderBy ? `&order_by=${orderBy}` : ""),
-    });
+        paginationOffset += paginationSize;
+        total = response.total;
+        previousTxCall = response.txs;
+        txs.push.apply(txs, response.txs);
+      } while (txs.length < total);
+      // The right condition should be nextPageToken != null but next_key isn't returned by the node for some reason
+    } catch (e) {
+      // Tx fetching failed, we return an empty array
+      return [];
+    }
+    return txs;
   }
 
-  private async getSendingTransactions(
+  private async fetchTransactions(
+    nodeUrl: string,
+    filterOn: string,
     address: string,
-    orderBy?: "ORDER_BY_DESC" | "ORDER_BY_ASC",
-  ) {
-    return await network({
+    options: {
+      "pagination.limit"?: number;
+      "pagination.offset"?: number;
+      "pagination.reverse"?: boolean;
+    },
+  ): Promise<{
+    txs: CosmosTx[];
+    nextPageToken: string | undefined;
+    total: number;
+  }> {
+    let serializedOptions = "";
+    for (const key of Object.keys(options)) {
+      serializedOptions += options[key] != null ? `&${key}=${options[key]}` : "";
```

Comment:
Typing says it's `number | undefined` but here you check for `null`

Reply:
We check for null and undefined (not !==)

I want to include both conditions

Reply:
is !options[key] is the same as options[key] != null ? 
I find is less disturbing to have !options[key] than options[key] != null 
`!=` is weird for me